SOF-2592: Add fitness for cross cutting concerns module

KvelaGorrrrnio opened the pull request:
>[!Note]
> The PR is based on the changes for #374.

Firstly, sorry for another big PR 🙏 

This PR adds a fitness function for ensuring that the cross-cutting-concerns module does not depend on other modules. The reason being that it is a special module where shared logic and concepts live and thus should not create transient dependencies for all the other modules.

The rule had some violations which led to the changes:

1. **Exceptions**
  - The base class for exceptions are moved to the cross-cutting-concerns-module.
  - The other exceptions has been moved to the modules that own the concept that thet describe.

2. **Mongo entity conversion**
  - The `MongoEntityConverter` has been split up and the individual converter methods are moved to the modules that own the entities that they convert from and to.

3. **Media**
  - All `Media` related code is moved to the sofie-ingest module as it builds on the media concept used in Sofie. We are most likely going to use a separate structure in Alba for asset management, where "playout items" and "media" is handled by the same asset system.

5. **Sofie data listening**
  - The `DataChangeService` and related code is moved to the sofie-ingest module as they all pertain the Sofie database.

6. **Entity test generation**
  - The `EntityTestFactory` is still placed in the rundown-execution module with a todo on splitting it up. This is not done in this PR as it has a scope where a separate PR is necessary to ensure that tests are not broken.
  - The `StatusMessage` part of the factory is moved into a `StatusMessageBuilder` as it was the only entity that the cross-cutting-concerns module is responsible for; and moving the `EntityTestFactory` to the cross-cutting-concerns, would it depend on all other modules.

7. **WebSocket event server uses only one event observer**
  - The `WebSocketEventServer` had ties to all other modules that produces events, so dependency direction is flipped, by introducing the event bus `TypedEventBus` that the individual event services can use to publish to the event server. This means that for most of the event services the observer interface is not necessary and are deleted. There is still one thing missing in that the WebSocketEventServer should be in the application layer. So it needs to be separated into transport and event logic. This will happen in a separate PR.

```mermaid
graph LR

typedEventServer["
  TypedEventServer
  [Service]

  Collects typed events and sends them the the websocket event server.
"]
typedEventEmitter["
  TypedEventEmitter
  [Interface]
"]
typedEventObserver["
  TypedEventObserver
  [Interface]
"]
typedEventBus["
    TypedEventBus
    [Service]
"]
rundownEventService["
    RundownEventService
    [Service]
    Builds and emits rundown events.
"]
statusMessageEventService["
    StatusMessageEventService
    [Service]
    
    Builds and emits status message events.
"]
XyzEventService["
    XyzEventService
    [Service]

    Builds and emits xyz events.
"]

typedEventObserver --> |"implemented by"| typedEventBus
typedEventEmitter --> |"implemented by"| typedEventBus



rundownEventService --> |"emits rundown events to"| typedEventEmitter
statusMessageEventService --> |"emits status message events to"| typedEventEmitter
XyzEventService --> |"emits xyz events to"| typedEventEmitter

typedEventServer --> |"subscribes to typed events from"| typedEventObserver
```

KvelaGorrrrnio committed:
ci: Adds fitness rule for outgoing dependencies of the cross-cutting-concerns module.

KvelaGorrrrnio committed:
refactor: Moves exceptions out into their respective modules.

KvelaGorrrrnio committed:
refactor: Splits up MongoEntityConverter into module-specialized versions.

KvelaGorrrrnio committed:
refactor: Moves all Media related code to sofie-ingest module.

KvelaGorrrrnio committed:
refactor: Moves DataChangeService to sofie-ingest module.

KvelaGorrrrnio committed:
refactor: Moves Device related code to the sofie-ingest module.

KvelaGorrrrnio committed:
test: Moves StatusMesssage test data building from entity-test-factory to status-message-builder.

KvelaGorrrrnio committed:
refactor: Adds TypedEventBus to make event services emit specialized events to the websocket event server.

This is done to avoid having the cross-cutting-concerns module depend on other context modules.

KvelaGorrrrnio committed:
chore: Adds todos to the fitness functions for module interaction about only having type imports between modules.

KvelaGorrrrnio committed:
chore: Adds allowed dependencies to dependency cruiser config.

Preparations for removing some of the negated forbidden rules.

KvelaGorrrrnio committed:
chore: Removes negated submodule fitness function rules from forbidden rules field in dependency cruiser config.

vushu commented on src/cross-cutting-concerns/domain/services/status-message-builder.ts:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  public withStatusCode(statusCode: StatusCode): StatusMessageBuilder {
+    return new StatusMessageBuilder({ ...this.statusMessage, statusCode })
```
Ah yeah using { ... } to return "self" I see :)

vushu commented on src/cross-cutting-concerns/infrastructure/services/web-socket-event-server.ts:
```diff
@@ -86,33 +61,7 @@ export class WebSocketEventServer implements EventServer {
-    this.playoutContentEventObserver.subscribeToPlayoutContentEvents((playoutContentEvent: PlayoutContentEvent) => {
-      webSocket.send(JSON.stringify(playoutContentEvent))
-    })
+    this.typedEventObserver.subscribeToTypedEvents((typedEvent: TypedEvent) => webSocket.send(JSON.stringify(typedEvent)))
```
Nice that events now can be put into a single `TypedEvent`.

vushu commented on src/rundown-execution/domain/exceptions/on-air-exception.ts:
```diff
@@ -1,5 +1,5 @@
-import { ErrorCode } from '../enums/error-code'
-import { Exception } from './exception'
+import { ErrorCode } from '../../../cross-cutting-concerns/domain/enums/error-code'
+import { Exception } from '../../../cross-cutting-concerns/domain/exceptions/exception'
```
So these exception are domain meaning in terms of automation and not according to bounded context?

KvelaGorrrrnio replied:
The `OnAirException` is a domain exception related to rundown execution. The `Exception` is a base class for all exceptions through all layers and is placed in the cross-cutting concerns domain module as it then is accessible from all other modules. It can be discussed whether that is the right approach :)
And I would like to remove the `ErrorCode` (in a separate PR) as it is redundant code for the exceptions.

vushu reviewed (approved):
Very cool, consider making a highlevel diagram that reflects how the modules interact with eachother using the event bus server. This would be a good supplement for better understanding atleast for me :D

LindvedKrvang commented on src/rundown-execution/domain/entities/test/entity-mock-factory.ts:
```diff
@@ -9,6 +9,7 @@ import { PartTimings } from '../../value-objects/part-timings'
 import { RundownMode } from '../../enums/rundown-mode'
 import { PlayoutContentType } from '../../enums/playout-content-type'
 
+// TODO: Get rid of it.
```
I disagree. Sometimes you do want mocks over actual instances.

KvelaGorrrrnio replied:
And I disagree with that :) Entities should not be mocked as they contain domain and business logic which is easily corrupted with mocks, so mocking entities create fragile tests and covers up regressions.

KvelaGorrrrnio committed:
Merge remote-tracking branch 'origin/AG-391/listenForIngestUpdatesForConfiugredQueues' into SOF-2592/add-fitness-for-cross-cutting-concerns-module

KvelaGorrrrnio committed:
fix: Corrects wrong path for importing Exceptions.

SOF-2592
After merge conflict some imports path to Exception was wrong due to
changed directory.

KvelaGorrrrnio committed:
chore: Changes Node engine requirement to >=16.20.

The lower limit is moved up to 16.20 as the mongodb driver requires this.
The upper limit is removed as the Buffer implementation in Node is no
longer broken on newer versions. Note that the Node version used by the
Dockerfile is unchanged.

KvelaGorrrrnio committed:
Merge branch 'develop' into SOF-2592/add-fitness-for-cross-cutting-concerns-module
